frame/merkle-mountain-range/src/tests.rs

```diff
@@ -347,27 +347,40 @@ fn should_verify() {
 }
 
 #[test]
-fn should_verify_batch_proof() {
+fn should_verify_batch_proofs() {
 	let _ = env_logger::try_init();
 
-	// Start off with chain initialisation and storing indexing data off-chain
-	// (MMR Leafs)
-	let mut ext = new_test_ext();
-	ext.execute_with(|| add_blocks(7));
-	ext.persist_offchain_overlay();
+	use itertools::Itertools;
+	let n = 13;
+	// generate all possible 2-leaf combinations for n = 13
+	let leaves_set: Vec<Vec<u64>> = (0..n).into_iter().combinations(2).collect();
 
-	// Try to generate proof now. This requires the offchain extensions to be present
-	// to retrieve full leaf data.
-	register_offchain_ext(&mut ext);
-	let (leaves, proof) = ext.execute_with(|| {
-		// when
-		crate::Pallet::<Test>::generate_batch_proof(vec![0, 4, 5]).unwrap()
-	});
+	leaves_set.iter().for_each(|leaves_subset| {
+		// Start off with chain initialisation and storing indexing data off-chain
+		// (MMR Leafs)
+		let mut ext = new_test_ext();
+		ext.execute_with(|| add_blocks(n as usize));
+		ext.persist_offchain_overlay();
 
-	ext.execute_with(|| {
-		add_blocks(7);
-		// then
-		assert_eq!(crate::Pallet::<Test>::verify_leaves(leaves, proof), Ok(()));
+		// Try to generate proof now. This requires the offchain extensions to be present
+		// to retrieve full leaf data.
+		register_offchain_ext(&mut ext);
```

Comment:
Good observation!
The reason I did the full loop was for the 7 blocks appended [after proof generation](https://github.com/paritytech/substrate/pull/11840/files#diff-bc8055a7324f264fb179940831634d3ea1812eef10b01728aecd9854f6e881f8L368), since I don't know how to revert these and start again from the same block (is there a clean way to do this?).
If I move everything else out of the loop and just keep adding blocks after every proof verification, the tests end up even longer (~9s) since we then have to generate & verify them on blockchains increasing depth by 7 for every verification: https://github.com/paritytech/substrate/pull/11840/commits/cda592d943c0faf042477b42d12d24e8dd5ecc29

Reply:
However: I _think_ the only reason those blocks are added to the test after proof generation is to test that proofs verify even if the chain has grown in the meantime (maybe @Wizdave97 remembers?).

If so, I don't think it's worth the performance impact, so if I skip that entirely in the loop and just test verification of a proof generated on an ancestor once (and for all other cases test it with the same state as available when generating the proof), I get much better performance: back down to ~0.5s with same testing range as before (https://github.com/paritytech/substrate/pull/11840/commits/2ca50555ccc4d24a5c1f8e5a3a89603ed115e6fc), and ~2s if I include a batch proof test for n > 1024 (https://github.com/paritytech/substrate/pull/11840/commits/9f3dfbd14ead8dcab29902940463cb6fd288e748)

Reply:
> However: I _think_ the only reason those blocks are added to the test after proof generation is to test that proofs verify even if the chain has grown in the meantime (maybe @Wizdave97 remembers?).

That’s right! I remember adding this for exactly this reason. And I agree that with the new test format you have, it’s not needed anymore.